vagrant/src/main/java/org/jclouds/vagrant/internal/VagrantNodeRegistry.java

```diff
@@ -73,36 +93,41 @@ public long getDelay(TimeUnit unit) {
          return unit.convert(expiryTime - timeSupplier.get(), TimeUnit.MILLISECONDS);
       }
    }
-   private Map<String, VagrantNode> nodes = new ConcurrentHashMap<String, VagrantNode>();
-   private DelayQueue<TerminatedNode> terminatedNodes = new DelayQueue<TerminatedNode>();
+
+   private final DelayQueue<TerminatedNode> terminatedNodes = new DelayQueue<TerminatedNode>();
+   private final Supplier<Map<String, VagrantNode>> nodes;
 
    private volatile long lastVacuumMs;
-   private Supplier<Long> timeSupplier;
+   private final Supplier<Long> timeSupplier;
 
    @Inject
-   VagrantNodeRegistry(@TimeStamp Supplier<Long> timeSupplier) {
+   VagrantNodeRegistry(@TimeStamp Supplier<Long> timeSupplier, Supplier<Collection<VagrantNode>> existingMachines) {
       this.timeSupplier = timeSupplier;
+      this.nodes = Suppliers.memoize(new ConcurrentWrapperSupplier(existingMachines));
```

Comment:
The `ConcurrentHashMap` is an implementation detail of the `VagrantNodeRegistry`. The wrapping supplier+memoize is just a mechanism to achieve lazy loading, not something that's expected to be provided by external wiring (so that we don't do IO in the constructor).
That's at least my take on it, don't mind making it pluggable if you still think that'd be useful.